Add support for playing storyboards beyond gameplay end time

Cublibre opened the pull request:
Closes [#11982](https://github.com/ppy/osu/issues/11982)

![storyboardtesting1](https://user-images.githubusercontent.com/40254686/114913770-9788c680-9def-11eb-83fe-aea2ee9ba7c1.gif)

### **Description** ###
Allows a user to skip a storyboard outro if there is one after the playable portion of the beatmap. 
Previously, the score was instantly shown after the last hit object regardless of the storyboard. Now, a skip overlay is shown and offers the option to skip to the end to display the score. This is disabled in multiplayer like skipping the intro.

- Adds a new `IStoryboardElementWithDuration` implemented by `IStoryboardSprite` and added an extension method `GetEndTime` to `IStoryboardElement` to calculate the end time of the storyboard.
- Skipping the outro, holding esc and waiting until the outro ends leads to `updateCompletionState` to avoid bypassing conditions checked before completion. 

### **Tests** ###
- Visual tests in `TestSceneStoryboardWithOutro`*
  - Skip outro with space
  - Skip outro with esc
  - Without skipping outro
  - Storyboard ends before score completion
  - Storyboard ends during fail animation
  - While toggling storyboard during the map
- ~~[ ] With cinema mod~~
  
*Unit tests can sometimes fail due to a bug described in #12426

### **Behavior Compared to Stable** ###

| Case                    | Original (stable)                                 | New behavior                                 |
| ----------------------- | ------------------------------------------------- | -------------------------------------------- |
| Player pressed skip     | Show score instantly                              | Show score instantly                         |
| Player exited using Esc | Show score instantly                              | Show score instantly                         |
| End of outro reached    | (unsure if with or without delay) shows the score | Show the score after `RESULTS_DISPLAY_DELAY` |

Co-Authored-By: @marlinabowring

Cublibre committed:
Expose the latest end time of storyboard elements

Co-authored-by: Marlina Bowring <marlina@umich.edu>

Cublibre committed:
Allow skipping storyboard outro

Reuses SkipOverlay by calculating the endtime of the storyboard and using that as a "start point". Upon skipping the outro the score is instantly shown.
When the end of the storyboard is reached the score screen automatically shows up. If the player holds ESC (pause) during the outro, the score is displayed

The storyboard endtime is calculated by getting the latest endtime of the storyboard's elements, or simply returning 0 if there is no storyboard.

Co-Authored-By: Marlina José <marlina@umich.edu>

frenzibyte commented on osu.Game/Screens/Play/GameplayClockContainer.cs:
```diff
@@ -296,5 +303,23 @@ public HardwareCorrectionOffsetClock(IClock source, bool processSource = true)
             }
         }
+
+        # region Storyboard outro logic
+
+        public IBindable<bool> HasStoryboardEnded => hasStoryboardEnded;
+
+        public bool HasTimeLeftInStoryboard => GameplayClock.CurrentTime <= StoryboardEndTime;
+
+        private readonly BindableBool hasStoryboardEnded = new BindableBool(true);
+
+        private void updateHasStoryboardEnded()
+        {
+            if (StoryboardEndTime == 0)
+                return;
+
+            hasStoryboardEnded.Value = GameplayClock.CurrentTime >= StoryboardEndTime;
+        }
+
+        # endregion
```
Better suited in `GameplayBeatmap` instead.

Cublibre replied:
I'm not sure I understand why. `GameplayBeatmap` doesn't have any reference to a storyboard, but `WorkingBeatmap` does. What is the difference between the two?

frenzibyte replied:
`GameplayClockContainer` shouldn't really have this logic in there, looks to be irrelevant from what it's doing. 

`GameplayBeatmap` is a component cached for DI made to provide information on the beatmap as well as tracking the current gameplay status, I see this logic better suited in that component, you can pass the storyboard along the beatmap there in the constructor. (FWIW, `BreakTracker`, a component that tracks whether the gameplay is in break time, should exist in `GameplayBeatmap` instead).

But now that you mention it, this logic could actually be in `DrawableStoryboard` instead, expose it from `DimmableStoryboard` for `Player` to access and there you have it.

Cublibre replied:
I see. Though, why define the logic in `DrawableStoryboard` instead of `DimmableStoryboard`? Additionally, how should I keep track of the time in `DrawableStoryboard`? I tried passing in the `GameplayClock` from `Player`, but then it gets messy since I have to change the constructor and `Storyboard.CreateDrawable()`. Is there a better approach?

frenzibyte replied:
`DimmableStoryboard` should do nothing but hiding, dimming, or blurring the storyboard, `DrawableStoryboard` is where this sort of logic should go into. (I also envision `LatestEndTime` to be moved there as well rather than in `Storyboard`, as elements like sample and video can't calculate their length without actually accessing the file and reading from there, but that's for another day)

As for the difficulty in passing `GameplayClock`, that's why [dependency injection](https://github.com/ppy/osu-framework/wiki/Dependency-Injection) is a thing, in fact, `DrawableStoryboard` already resolves `GameplayClock` and respects it:
https://github.com/ppy/osu/blob/215ad780454bb0f9fd2a238e1b3334e7a0ee0c61/osu.Game/Storyboards/Drawables/DrawableStoryboard.cs#L62-L66

So you can keep track of time by [`Time.Current`](https://github.com/ppy/osu-framework/blob/99786238a02e0d6b69da86dd52e5506ee8ec0566/osu.Framework/Graphics/Transforms/Transformable.cs#L25-L28), which is set to `GameplayClock` in here.

frenzibyte commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -285,6 +291,9 @@ private void load(AudioManager audio, OsuConfigManager config, OsuGameBase game)
             ScoreProcessor.HasCompleted.ValueChanged += updateCompletionState;
             HealthProcessor.Failed += onFail;
 
+            // Keep track of whether the storyboard ended after the playable portion
```
```suggestion
```
unnecessary.

frenzibyte commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -106,6 +108,8 @@ public abstract class Player : ScreenWithBeatmapBackground, ISamplePlaybackDisab
 
         private SkipOverlay skipOverlay;
```
rename to `skipIntroOverlay` then.

frenzibyte commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -190,6 +194,8 @@ private void load(AudioManager audio, OsuConfigManager config, OsuGameBase game)
 
             mouseWheelDisabled = config.GetBindable<bool>(OsuSetting.MouseDisableWheel);
 
+            storyboardEnabled = config.GetBindable<bool>(OsuSetting.ShowStoryboard);
```
This shouldn't read from this setting solely, I'm pretty sure that'll break cinema mods from handling storyboard outros if this is disabled, since cinema mods ignore user settings.

At least use `DimmableStoryboard.ContentDisplayed`.

frenzibyte commented on osu.Game/Screens/Play/GameplayClockContainer.cs:
```diff
@@ -296,5 +303,23 @@ public HardwareCorrectionOffsetClock(IClock source, bool processSource = true)
+
+        public IBindable<bool> HasStoryboardEnded => hasStoryboardEnded;
+
+        public bool HasTimeLeftInStoryboard => GameplayClock.CurrentTime <= StoryboardEndTime;
```
unnecessary? should just read from `HasStoryboardEnabled.Value`

Cublibre replied:
Is `DimmableStoryboard.ContentDisplayed` dynamically updated after the end of the storyboard is reached? I'm not sure what's the best way to check if the storyboard has ended.

frenzibyte replied:
instead of `HasTimeLeftInStoryboard`, use `!HasStoryboardEnded.Value`, that's what I mean.

Cublibre committed:
Merge branch 'master' into play-storyboard-outro

Cublibre commented:
I updated the branch so it includes `TestSceneEditorSelection`

frenzibyte commented on osu.Game/Screens/Play/PlayerConfiguration.cs:
```diff
@@ -24,5 +24,10 @@ public class PlayerConfiguration
         /// Whether the player should be allowed to skip the intro, advancing to the start of gameplay.
         /// </summary>
         public bool AllowSkippingIntro { get; set; } = true;
+
+        /// <summary>
+        /// Whether the player should be allowed to skip the outro, advancing to the end of a storyboard.
+        /// </summary>
+        public bool AllowSkippingOutro { get; set; } = true;
```
I would just bundle them all in one `AllowSkipping`, the point is whether or not the player is allowed to skip in time if they desire.

Cublibre committed:
Add IHasDuration interface with EndTime for storyboard elements to implement

Cublibre commented:
Should the skip overlay be shown during the outro if Cinema mod is enabled? I noticed that on stable the skip overlay is not shown at all.

peppy commented:
Probably not, but that would be a separate issue / PR (along with hiding it during the intro).

Cublibre committed:
Move storyboard outro logic to DrawableStoryboard

Cublibre committed:
Clean up skipOutroOverlay if skipping is disabled

frenzibyte commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -285,6 +287,10 @@ private void load(AudioManager audio, OsuConfigManager config, OsuGameBase game)
 
+            // Keep track of whether the storyboard ended after the playable portion
+            if (DimmableStoryboard.HasStoryboardEnded != null)
+                DimmableStoryboard.HasStoryboardEnded.ValueChanged += updateCompletionState;
```
Wouldn't this cause `PrepareScoreForResultAsync` to be called twice? also what happens if the storyboard is less than gameplay, sounds like there should be further logic in `updateCompletionState` to consider completion when both score processor and storyboard are done.

bdach replied:
To me it feels like gameplay completion and storyboard completion should not be using this same code path as currently written. Both conditions need to be checked before proceeding with exit.

Cublibre replied:
Is this a better approach?

```cs
if (DimmableStoryboard.HasStoryboardEnded != null)
{
    DimmableStoryboard.HasStoryboardEnded.ValueChanged += (storyboardEnded) =>
    {
        if (storyboardEnded.NewValue && ScoreProcessor.HasCompleted.Value)
            scheduleCompletion();
    };
}
```

frenzibyte commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -611,6 +634,15 @@ private void updateCompletionState(ValueChangedEvent<bool> completionState)
                 return score.ScoreInfo;
             });
 
+            var storyboardHasOutro = DimmableStoryboard.ContentDisplayed && (!DimmableStoryboard.HasStoryboardEnded?.Value ?? false);
+
+            if (storyboardHasOutro)
+            {
+                skipOutroOverlay.Show();
+                completionProgressDelegate = null;
+                return;
+            }
```
```suggestion
            if (DimmableStoryboard.ContentDisplayed &&
                DimmableStoryboard.HasStoryboardEnded?.Value == false)
            {
                skipOutroOverlay.Show();
                return;
            }
```

bdach replied:
Disagree with inlining the conditional. Variable is fine. Just discarding the delegate without cancellation is extremely dicey, though.

frenzibyte replied:
The delegate should not be non-null at this point, that's why I removed it in my suggestion, to avoid confusion.

bdach commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -285,6 +287,10 @@ private void load(AudioManager audio, OsuConfigManager config, OsuGameBase game)
             HealthProcessor.Failed += onFail;
 
+            // Keep track of whether the storyboard ended after the playable portion
+            if (DimmableStoryboard.HasStoryboardEnded != null)
+                DimmableStoryboard.HasStoryboardEnded.ValueChanged += updateCompletionState;
```
I think it would suffice to have a separate `HasStoryboardEnded` bindable in `DimmableStoryboard`, defaulting to `true`, and just bind to it normally on storyboard load, instead of directly proxying as done currently. Would also spare the null check here, too.

frenzibyte replied:
One thing is that I wasn't sure about letting `DimmableStoryboard` expose bindables from `DrawableStoryboard` by creating each and bind on storyboard creation, but instead leave each class do its own thing.

I guess your solution is fine for the fact that `DimmableStoryboard` is still a storyboard-related class.

bdach commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -523,6 +538,14 @@ protected void PerformExit(bool showDialogFirst)
                     return;
                 }
+
+                // show the score if in storyboard outro (score has been set)
+                bool scoreReady = prepareScoreForDisplayTask != null && prepareScoreForDisplayTask.IsCompleted;
+
+                if (scoreReady)
+                {
+                    scheduleCompletion();
+                }
```
I'm not even sure what the original intent was here, looking at `scheduleCompletion()`'s body. It's explicitly written in a way that should automatically handle task completion.

bdach commented on osu.Game/Storyboards/Storyboard.cs:
```diff
@@ -36,6 +36,12 @@ public class Storyboard
+        /// Across all layers, find the latest point in time that a storyboard element ends at.
+        /// Will return null if there are no elements.
+        /// </summary>
+        public double? LatestEventTime => Layers.SelectMany(l => l.Elements.OfType<IStoryboardElementHasDuration>()).OrderByDescending(e => e.EndTime).FirstOrDefault()?.EndTime;
```
Extension method (following the precedent of hitobjects) seems fine to me.

bdach commented on osu.Game/Storyboards/IStoryboardElementHasDuration.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace osu.Game.Storyboards
+{
+    public interface IStoryboardElementHasDuration
+    {
+        double EndTime { get; }
+    }
```
Needs xmldoc.

bdach commented on osu.Game/Storyboards/Drawables/DrawableStoryboard.cs:
```diff
@@ -82,5 +83,27 @@ private void updateLayerVisibility()
+        /// </summary>
+        public IBindable<bool> HasStoryboardEnded => hasStoryboardEnded;
+
+        private readonly BindableBool hasStoryboardEnded = new BindableBool(true);
+
+        private void updateHasStoryboardEnded()
+        {
+            if (Storyboard.LatestEventTime == null)
+                return;
+
+            var time = Clock.CurrentTime;
+            hasStoryboardEnded.Value = time >= Storyboard.LatestEventTime;
+        }
```
Would probably shorten to

```csharp
hasStoryboardEnded.Value = Storyboard.LatestEventTime == null
	|| Time.Current >= Storyboard.LatestEventTime;
```

bdach commented on osu.Game/Storyboards/Storyboard.cs:
```diff
@@ -36,6 +36,16 @@ public class Storyboard
+        /// </summary>
+        /// <remarks>
+        /// This iterates all elements and as such should be used sparingly or stored locally.
+        /// Videos and samples return StartTime as their EndTIme.
```
I actually would like this comment back... Clarifies up the kink about videos/samples.

frenzibyte commented on osu.Game/Screens/Play/GameplayClockContainer.cs:
```diff
@@ -228,13 +228,20 @@ public void StopUsingBeatmapClock()
             adjustableClock.ChangeSource(track);
         }
 
+        /// <summary>
+        /// Gets the endtime of the last element in the storyboard in ms, or start time of the last hitobject if there's no storyboard.
+        /// </summary>
+        public double StoryboardEndTime => beatmap.Storyboard.LatestEventTime ?? 0;
```
logic was different at the time of posting, but this shouldn't be here in `GameplayClockContainer`, just read from the beatmap's storyboard directly.

Cublibre committed:
Read StoryboardEndTime directly from Beatmap

Cublibre committed:
Add remarks back to LatestEventTime

frenzibyte commented on osu.Game.Tests/Visual/Gameplay/TestSceneStoryboardWithOutro.cs:
```diff
@@ -0,0 +1,113 @@
+            {
+            }
+
+            protected override Task ImportScore(Score score)
+            {
+                // avoid database errors from trying to store the score
+                return Task.CompletedTask;
+            }
```
```
[database] 2021-04-17 16:34:08 [verbose]: [?????] Beginning import...
[database] 2021-04-17 16:34:10 [error]: Failed executing DbCommand (8ms) [Parameters=[@p10='?', @p11='?', @p12='?', @p13='?', @p14='?', @p15='?', @p16='?', @p17='?', @p18='?', @p19='?', @p20='?', @p21='?', @p22='?', @p23='?', @p24='?', @p25='?', @p26='?', @p27='?', @p28='?', @p29='?', @p30='?', @p31='?', @p32='?', @p33='?', @p34='?' (Size = 6), @p35='?'], CommandType='Text', CommandTimeout='10']
[database] 2021-04-17 16:34:10 [error]: INSERT INTO "BeatmapInfo" ("AudioLeadIn", "BPM", "BaseDifficultyID", "BeatDivisor", "BeatmapSetInfoID", "Countdown", "DistanceSpacing", "EpilepsyWarning", "GridSize", "Hash", "Hidden", "Length", "LetterboxInBreaks", "MD5Hash", "MetadataID", "OnlineBeatmapID", "Path", "RulesetID", "SpecialStyle", "StackLeniency", "StarDifficulty", "Status", "StoredBookmarks", "TimelineZoom", "Version", "WidescreenStoryboard")
[database] 2021-04-17 16:34:10 [error]: VALUES (@p10, @p11, @p12, @p13, @p14, @p15, @p16, @p17, @p18, @p19, @p20, @p21, @p22, @p23, @p24, @p25, @p26, @p27, @p28, @p29, @p30, @p31, @p32, @p33, @p34, @p35);
[database] 2021-04-17 16:34:10 [error]: SELECT "ID"
[database] 2021-04-17 16:34:10 [error]: FROM "BeatmapInfo"
[database] 2021-04-17 16:34:10 [error]: WHERE changes() = 1 AND "ID" = last_insert_rowid();
[database] 2021-04-17 16:34:10 [error]: An exception occurred in the database while saving changes for context type 'osu.Game.Database.OsuDbContext'.
[database] 2021-04-17 16:34:10 [error]: Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while updating the entries. See the inner exception for details.
[database] 2021-04-17 16:34:10 [error]: ---> Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 19: 'FOREIGN KEY constraint failed'.
[database] 2021-04-17 16:34:10 [error]: at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)
[database] 2021-04-17 16:34:10 [error]: at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
[database] 2021-04-17 16:34:10 [error]: at Microsoft.Data.Sqlite.SqliteCommand.ExecuteDbDataReader(CommandBehavior behavior)
[database] 2021-04-17 16:34:10 [error]: at System.Data.Common.DbCommand.ExecuteReader()
[database] 2021-04-17 16:34:10 [error]: at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
[database] 2021-04-17 16:34:10 [error]: at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
[database] 2021-04-17 16:34:10 [error]: at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
[database] 2021-04-17 16:34:10 [error]: --- End of inner exception stack trace ---
```

I believe it's due to importing a score info of a beatmap which doesn't exist in the database, that's the main reason I've suggested moving this at `TestPlayer`, as mostly all of them contain programmatically written beatmaps.

And in the end the scores will be imported to your actual database when running this in visual tests, and I don't think that's really a wanted behaviour. If anything, it could be a toggleable property if some tests want to test the score import process.

bdach replied:
> I believe it's due to importing a score info of a beatmap which doesn't exist in the database

that doesn't mean that this will always be the case. some test cases want to import actual maps and import actual scores. i'd rather this behaviour of crippling imports be opt-out and not opt-in. more explicit, less of silent "things are disabled in tests by default" hackery.

> And in the end the scores will be imported to your actual database when running this in visual tests

not necessarily? would recommend checking out the `bin/Debug/net5.0/` directory in a test project. in particular you should be seeing separate `client.db` files.

Cublibre replied:
I got FOREIGN KEY constraint failed without the suppression. I think it's because the game was trying to import the fake beatmap I created, while other tests use the `TestBeatmap` to completion or don't play out a fake beatmap to the score screen (`TestSceneCompletionCancellation`)

Here's part of the stack trace:

```
[database] 2021-04-17 16:42:10 [error]: Failed executing DbCommand (0ms) [Parameters=[@p10='?', @p11='?', @p12='?', @p13='?', @p14='?', @p15='?', @p16='?', @p17='?', @p18='?', @p19=
'?', @p20='?', @p21='?', @p22='?', @p23='?', @p24='?', @p25='?', @p26='?', @p27='?', @p28='?', @p29='?', @p30='?', @p31='?', @p32='?', @p33='?', @p34='?' (Size = 6), @p35='?'], Comm
andType='Text', CommandTimeout='10']
[database] 2021-04-17 16:42:10 [error]: INSERT INTO "BeatmapInfo" ("AudioLeadIn", "BPM", "BaseDifficultyID", "BeatDivisor", "BeatmapSetInfoID", "Countdown", "DistanceSpacing", "Epil
epsyWarning", "GridSize", "Hash", "Hidden", "Length", "LetterboxInBreaks", "MD5Hash", "MetadataID", "OnlineBeatmapID", "Path", "RulesetID", "SpecialStyle", "StackLeniency", "StarDif
ficulty", "Status", "StoredBookmarks", "TimelineZoom", "Version", "WidescreenStoryboard")
[database] 2021-04-17 16:42:10 [error]: VALUES (@p10, @p11, @p12, @p13, @p14, @p15, @p16, @p17, @p18, @p19, @p20, @p21, @p22, @p23, @p24, @p25, @p26, @p27, @p28, @p29, @p30, @p31, @
p32, @p33, @p34, @p35);
[database] 2021-04-17 16:42:10 [error]: SELECT "ID"
[database] 2021-04-17 16:42:10 [error]: FROM "BeatmapInfo"
[database] 2021-04-17 16:42:10 [error]: WHERE changes() = 1 AND "ID" = last_insert_rowid();
[database] 2021-04-17 16:42:10 [error]: An exception occurred in the database while saving changes for context type 'osu.Game.Database.OsuDbContext'.
[database] 2021-04-17 16:42:10 [error]: Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while updating the entries. See the inner exception for details.
[database] 2021-04-17 16:42:10 [error]: ---> Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 19: 'FOREIGN KEY constraint failed'.
[database] 2021-04-17 16:42:10 [error]: at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)
[database] 2021-04-17 16:42:10 [error]: at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
[database] 2021-04-17 16:42:10 [error]: at Microsoft.Data.Sqlite.SqliteCommand.ExecuteDbDataReader(CommandBehavior behavior)
[database] 2021-04-17 16:42:10 [error]: at System.Data.Common.DbCommand.ExecuteReader()
[database] 2021-04-17 16:42:10 [error]: at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod,
IReadOnlyDictionary`2 parameterValues)
[database] 2021-04-17 16:42:10 [error]: at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 par
ameterValues)
[database] 2021-04-17 16:42:10 [error]: at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
[database] 2021-04-17 16:42:10 [error]: --- End of inner exception stack trace ---
```

frenzibyte replied:
> > I believe it's due to importing a score info of a beatmap which doesn't exist in the database
> 
> that doesn't mean that this will always be the case. some test cases want to import actual maps and import actual scores. i'd rather this behaviour of crippling imports be opt-in and not opt-out. more explicit, less of silent "things are disabled in tests by default" hackery.

I guess you have a point on this one.

> > And in the end the scores will be imported to your actual database when running this in visual tests
> 
> not necessarily? would recommend checking out the `bin/Debug/net5.0/` directory in a test project. in particular you should be seeing separate `client.db` files for particular test scene executions.

I believe the separate `client.db` files are for NUnit test executions, not visual tests browser:
![CleanShot 2021-04-17 at 20 07 54](https://user-images.githubusercontent.com/22781491/115120935-c56d3700-9fb8-11eb-9607-05c184df31fc.png)

Cublibre committed:
Update TestSceneStoryboardWithOutro.cs

- Construct storyboard in CreateWorkingBeatmap()
- Use GameplayClockContainer.GameplayClock.CurrentTime
- Remove unnecessary lines

Cublibre commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -250,7 +246,7 @@ private void load(AudioManager audio, OsuConfigManager config, OsuGameBase game)
                 HUDOverlay.ShowHud.Disabled = true;
                 BreakOverlay.Hide();
-                skipOverlay.Hide();
+                skipIntroOverlay.Hide();
```
Should I keep the hide in `createOverlayComponents()`. or is there a better way to make `skipOutroOverlay` hidden at the beginning?

frenzibyte replied:
As per above, should call `Hide()` here as well, regardless of whether it was hidden before.

Cublibre committed:
Use GetEndTime() to get storyboard endtime

Cublibre committed:
Initialize skipOutroOverlay with alpha 0, other small changes

Cublibre committed:
Show score after the end of the storyboard after it was toggled

Cublibre committed:
HasStoryboardEnded doesn't trigger updateCompletionState()

Scores won't be shown prematurely if the storyboard ends before the playable portion of the beatmap.

Cublibre committed:
Added a test for toggling the storyboard after the map is loaded

Cublibre committed:
Remove redundant string interpolation

frenzibyte committed:
Schedule completion when storyboard has actually ended

frenzibyte committed:
Add further xmldoc

frenzibyte committed:
Merge branch 'master' into play-storyboard-outro

frenzibyte committed:
Remove new line between skip overlay fields

frenzibyte commented on osu.Game/Storyboards/IStoryboardElementHasDuration.cs:
```diff
@@ -0,0 +1,13 @@
+    /// <summary>
+    /// A StoryboardElement that ends at a different time than its start time.
+    /// </summary>
+    public interface IStoryboardElementHasDuration
```
I think it would be far better if this becomes a "with duration" interface instead:
```csharp
public interface IStoryboardElementWithDuration : IStoryboardElement
{
     double EndTime { get; }
}
```

bdach replied:
Also now that I think about it, it's a bit weird that this interface has an `EndTime` member but no `Duration`. Especially that it's a "storyboard element with duration".

Cublibre replied:
Should I move `EndTime` back to `IStoryboardElement`?

frenzibyte replied:
no it's to add another member `Duration` just like in `IHasDuration`.

frenzibyte reviewed (changes requested):
One major note and few remianing points, otherwise this looks fine to me as an initial step.

Cublibre committed:
Rename IStoryboardElementHasDuration, remove unnecessary step in tests, add Duration field

frenzibyte commented on osu.Game/Screens/Play/DimmableStoryboard.cs:
```diff
@@ -19,6 +20,14 @@ public class DimmableStoryboard : UserDimContainer
+        /// <remarks>
+        /// This is true by default in here, until an actual drawable storyboard is loaded, in which case it'll bind to it.
+        /// </remarks>
+        public IBindable<bool> HasStoryboardEnded = new BindableBool(true);
```
To reiterate on https://github.com/ppy/osu/pull/12432#discussion_r615210904, this is how I thought about having it https://github.com/frenzibyte/osu/commit/425d6b42ffcf52c2faee8fc0167d1ae73d011df1, while the current solution pushed here is simpler, the changes I made in the linked commit keeps `DimmableStoryboard` from having to expose things in `DrawableStoryboard` and initialize with a sane default.

I'm fine either way, but going to leave this as a comment.

bdach replied:
@frenzibyte your version is a bit weird (private event for some reason with a separate binding method, instead of just having the event public and then invoking it at the right time), and I honestly prefer the variant as written here. Makes sense to me to have `HasStoryboardEnded` on both `DrawableStoryboard` and this.

frenzibyte replied:
> @frenzibyte your version is a bit weird (private event for some reason with a separate binding method, instead of just having the event public and then invoking it at the right time)

actually yeah could make it public and assign the event at `createUnderlayComponents` right after creating the dimmable before returning it for load, didn't think about it.

bdach commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -281,6 +284,12 @@ private void load(AudioManager audio, OsuConfigManager config, OsuGameBase game)
+            DimmableStoryboard.HasStoryboardEnded.ValueChanged += _ =>
+            {
+                if (ScoreProcessor.HasCompleted.Value)
+                    scheduleCompletion();
```
Guard checking all completion conditions at the start of `updateCompletionState()` sounds fine.

marlinabowring replied:
How should we cause the player to fail?

frenzibyte replied:
Make a private custom mod that causes the player to fail immediately and not restart on fail (check `IApplicableFailOverride`).

marlinabowring replied:
How do we specify failure before the first hit circle?

frenzibyte replied:
Hmm, that's actually not a good solution, disregard it. Instead: 
 1. override `HasCustomSteps` to true, and call `CreateTest(null)` on all test cases. (that needs to happen any way you slice it)
 2. make a private field in the test scene holding the test case's own fail conditions, reset it on `SetUp` and set it on the test case you want the player to fail on, by: 
 ```csharp
 CreateTest(() => 
 {
     AddStep("fail on first judgement", () => failConditions = (_, __) => true); 
 }
 ```
 
 3. then pass that fail conditions field to `OutroPlayer` and subscribe to `HealthProcessor.FailConditions` with the given conditions.

Cublibre replied:
Why is it bad to just alter the `FailConditions` on one testcase? 

We found that we actually don't need to fail immediately- we're having trouble getting the storyboard to end during the fail animation while still having time for the skip overlay to be clicked in `TestSkipOutro`. We want to change the test to check if the overlay is simply visible instead, but the `skipOutroOverlay` is private so we can't access it with `State.Value`. Is there some way to check if some overlay is visible?

frenzibyte replied:
> Why is it bad to just alter the `FailConditions` on one testcase?
>
> We found that we actually don't need to fail immediately-

I'm not sure what you mean, do you have an alternative to what I proposed? if so, post it here.

> we're having trouble getting the storyboard to end during the fail animation while still having time for the skip overlay to be clicked in `TestSkipOutro`. We want to change the test to check if the overlay is simply visible instead, but the `skipOutroOverlay` is private so we can't access it with `State.Value`. Is there some way to check if some overlay is visible?

Usually, it's retrieved by using the testing extension `ChildrenOfType<T>`, but since there are two skip overlays I would say just expose it (`protected` at base and `public` at test class)

Cublibre replied:
> I'm not sure what you mean, do you have an alternative to what I proposed? if so, post it here.

I meant we didn't need the player to fail immediately during the test case, in fact doing that would make the storyboard end further after the fail animation (which doesn't test what we want.) So it's fine for the player to fail after the first judgement.

If the storyboard lasts 2000ms, then it won't end during the fail animation. We found that if the storyboard is 1300ms, then it will end during the fail animation but then `TestSkipOutro` is broken because the storyboard ends too quickly after the last hitobject for the test to click on the skip overlay though it does show up briefly. 

> Usually, it's retrieved by using the testing extension `ChildrenOfType<T>`, but since there are two skip overlays I would say just expose it (`protected` at base and `public` at test class)

It seems like I still can't access `State.Value` because it's inside a private class `FadeContainer`. Should I use `ChildrenOfType<T>`?

Cublibre replied:
Checking if the skip overlay is present seems to work, but then it can't test if clicking on it shows the score.

Is there some way I could change the storyboard duration before each test? Otherwise it seems like I'll have to make a new test scene for the fail overlay.

frenzibyte replied:
As I said, you'll have to manually load the player for each test case to achieve that, check the steps in https://github.com/ppy/osu/pull/12432#discussion_r615468528, add another field `currentStoryboard` and set it in the same place as the fail conditions:
```csharp
CreateTest(() =>
{
    AddStep("set storyboard" () => currentStoryboard = ...); 
    AddStep("fail on first judgement", () => currentFailConditions = (_, __) => true); 
});
```

then pass `currentStoryboard` in `CreateWorkingBeatmap` and there you have it.

Cublibre commented:
> As a good step, scheduleCompletion should never be called from anywhere else except for updateCompletionState.

I think it's fine for the skip outro overlay to call `scheduleCompletion` right? Because the overlay is shown after  `prepareScoreForDisplay` is non-null.

frenzibyte commented:
I don't think so, it would still bypass certain conditions in `updateCompletionState`. Any reason not to call it there?

Cublibre committed:
Storyboard completion calls updateCompletionState

- If the storyboard ends after the beatmap, call updateCompletionState as if the score processor has completed at that time. (completionProgressDelegate is null here since earlier when the score processor actually completed, updateCompletionState returned after showing the skip overlay.)

- If the storyboard ends before the beatmap does, updateCompletionState simply returns and waits until the score processor is completed.

- If the storyboard and beatmap end at the exact same time, make sure updateCompletionState() is called only once by the score processor completion.

Co-Authored-By: Marlina José <marlina@umich.edu>

Cublibre committed:
Added tests

New tests:
- storyboard ending during the failing animation
- showResults = false
- storyboard ending before score processor completion

Co-Authored-By: Marlina José <marlina@umich.edu>

Cublibre commented:
> I don't think so, it would still bypass certain conditions in `updateCompletionState`. Any reason not to call it there?

If the skip overlay calls `updateCompletionState`, then the score is shown after a delay (`RESULTS_DISPLAY_DELAY=1000`), which doesn't match stable where it seems to be shown instantly.

frenzibyte commented:
I was gonna say you should be able to add a parameter in `updateCompletionState` for removing the delay (false by default), could just do this and it should work:
```csharp
using (BeginDelayedSequence(immediateTransition ? RESULTS_DISPLAY_DELAY : 0))
    scheduleCompletion();
```

Cublibre commented:
I found the issue: when `updateCompletionState` is called by the skip overlay, completion is blocked by the outro check, so I checked for `immediateTransition` before that and called `scheduleCompletion`.
```cs
if (immediateTransition)
    scheduleCompletion();

var storyboardHasOutro = DimmableStoryboard.ContentDisplayed && !DimmableStoryboard.HasStoryboardEnded.Value;
if (storyboardHasOutro)
{
    skipOutroOverlay.Show();
    return;
}

using (BeginDelayedSequence(RESULTS_DISPLAY_DELAY))
    scheduleCompletion();
```

frenzibyte commented:
If that's how you'll use it then the parameter should be renamed to something like `skipGameplayOutro`

Cublibre commented:
I wonder, what conditions could be bypassed if the skip overlay called `scheduleCompletion()` directly? Aren't all the conditions (`HasFailed`, `ShowResults`) be already checked in `updateCompletionState` before the skip overlay is shown?

frenzibyte commented:
Still doesn't change my mind, `updateCompletionState` should be the only method responsible for calling `scheduleCompletion`, regardless of whether the order of things show that the outro skip overlay can bypass checks just fine.

Cublibre committed:
Skip outro overlay and PerformExit() call updateCompletionState() instead of scheduleCompletion()

Co-Authored-By: Marlina José <marlina@umich.edu>

Cublibre commented:
Alright, I also made `PerformExit()` call `updateCompletionState` the same way the skip overlay does for when the player holds esc during the outro.

frenzibyte commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -581,7 +581,7 @@ public void Restart()
         private Task<ScoreInfo> prepareScoreForDisplayTask;
 
-        private void updateCompletionState(ValueChangedEvent<bool> completionState)
+        private void updateCompletionState(ValueChangedEvent<bool> completionState, bool skipStoryboardOutro = false)
```
`completionState` should no longer exist now, use `ScoreProcessor.HasCompleted.Value` instead

Cublibre committed:
Remove completionState parameter

peppy committed:
Merge branch 'master' into play-storyboard-outro

peppy committed:
Implement Duration via the interface

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -355,10 +363,15 @@ private Drawable createOverlayComponents(WorkingBeatmap working)
                         Origin = Anchor.Centre
                     },
-                    skipOverlay = new SkipOverlay(DrawableRuleset.GameplayStartTime)
+                    skipIntroOverlay = new SkipOverlay(DrawableRuleset.GameplayStartTime)
                     {
                         RequestSkip = performUserRequestedSkip
                     },
+                    skipOutroOverlay = new SkipOverlay(Beatmap.Value.Storyboard.LatestEventTime ?? 0)
+                    {
+                        RequestSkip = () => updateCompletionState(true),
+                        Alpha = 0
+                    },
```
While I think this is fine for now for simplicity, there may be a point in the future where this needs to be refactored into a single overlay able to handle multiple skip operations (for instance, when implementing #12493).

frenzibyte replied:
Yeah, not to mention that the current way would break when rewinding before the storyboard outro part, which can be fixed easily with a single skip overlay (potentially by using the existing `PeriodTracker` to tell whether a period can be skipped).

peppy replied:
Rather than a reply to an inline comment, that seems like something you should post in a review? It is an oversight that blocks this from being usable.

frenzibyte replied:
Right, I thought it's not easily solvable and something to resolve in a follow-up PR or something, but now that I think about it, it can be solved and does currently block it from being usable, will do.

peppy committed:
Specify explicit primitive type

peppy committed:
Fix `AllowGameplayOverlays` potentially not working for outro skip overlay

peppy committed:
Add xmldoc for `updateCompletionState`

peppy commented on osu.Game/Storyboards/Storyboard.cs:
```diff
@@ -36,6 +36,16 @@ public class Storyboard
+        /// Will return null if there are no elements.
+        /// </summary>
+        /// <remarks>
+        /// This iterates all elements and as such should be used sparingly or stored locally.
```
You included this note but didn't follow it yourself 😅

peppy committed:
Cache last event time value to avoid super expensive LINQ

peppy commented:
I've made some changes to get this into a state I'm happy with merging. Would appreciate one more review though.

frenzibyte reviewed (approved):
Checked this again, looks good to me.

peppy reviewed (changes requested):
Self blocking due to failure on rewind.

frenzibyte reviewed (changes requested):
As been mentioned above, the outro skip overlay should be hidden when rewinding before the gameplay end (should hide in the `!completionState.NewValue` branch in `updateCompletionState`)

Cublibre committed:
Hide and disable skip outro overlay on rewind

Cublibre committed:
Add visual test for rewinding

Cublibre commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -564,17 +582,23 @@ public void Restart()
                 completionProgressDelegate?.Cancel();
                 completionProgressDelegate = null;
                 ValidForResume = true;
+                skipOutroOverlay.Hide();
```
This was all I did to disable the overlay upon rewinding, but after reaching the outro again, the overlay fades out immediately unlike when it is first shown. I think this is caused by the logic in `SkipOverlay.cs` that keeps track of the progress since `displayTime` isn't updated after the overlay is first loaded.

https://github.com/ppy/osu/blob/68807a90ec4f0251252ba728d8d12d1be5201655/osu.Game/Screens/Play/SkipOverlay.cs#L115-L120

peppy replied:
I'm confused. You're saying it still doesn't work correctly, but the test you added passes. Do you have another test which fails?

peppy replied:
I've adjusted the test and fixed the underlying issue (state not being passed on to `FadeContainer`, which is the one handling the "fade on user movement" part).

peppy committed:
Actually check for correct state of fade content in rewind test

peppy committed:
Fix `SkipOverlay`'s `FadeContent` not getting correct state from parent

peppy committed:
Merge branch 'master' into play-storyboard-outro
